File: crates/web-lua-core/src/web/tab.rs
use crate::state::HostState;
use piccolo::{Context, Table};
use std::cell::RefCell;
use std::rc::Rc;

pub(crate) fn register<'a>(ctx: Context<'a>, host_state: Rc<RefCell<HostState>>) -> Table<'a> {
    let _tab_table = Table::new(&ctx);

    let tab_table = Table::new(&ctx);
    lua_api!(ctx, tab_table,
    name: "query",
    action: "tab_query",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Query Chrome tabs matching given criteria.",
    params: [
    query_info: "table", optional, "Query filter: active, currentWindow, url, etc.",
    ],
    returns: "table" => "Array of matching tab objects",
    );
    lua_api!(ctx, tab_table,
    name: "create",
    action: "tab_create",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Create a new tab.",
    params: [
    create_properties: "table", optional, "URL, windowId, active, etc.",
    ],
    returns: "table" => "Created tab object",
    );
    lua_api!(ctx, tab_table,
    name: "activate",
    action: "tab_activate",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Activate (focus) a tab.",
    params: [
    tab_id: "number", required, "Tab ID to activate",
    ],
    returns: "boolean" => "Whether activation succeeded",
    );
    lua_api!(ctx, tab_table,
    name: "close",
    action: "tab_close",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Close a tab.",
    params: [
    tab_id: "number", required, "Tab ID to close",
    ],
    returns: "boolean" => "Whether close succeeded",
    );
    lua_api!(ctx, tab_table,
    name: "execute_script",
    action: "tab_execute_script",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Execute JavaScript in a target tab.",
    params: [
    tab_id: "number", required, "Target tab ID",
    script: "string | table", required, "Script code or injection details",
    ],
    returns: "table" => "Injection results",
    );
    lua_api!(ctx, tab_table,
        name: "click",
        action: "tab_click",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Click an element by refId in the target tab.",
        params: [
            tab_id: "number", required, "Target tab ID",
            ref_id: "number", required, "Element refId from snapshot",
        ],
        returns: "boolean" => "Whether the click succeeded",
    );
    lua_api!(ctx, tab_table,
    name: "fill",
    action: "tab_fill",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Fill an input element by refId in the target tab.",
    params: [
    tab_id: "number", required, "Target tab ID",
    ref_id: "number", required, "Element refId from snapshot",
    value: "string", required, "Text to fill",
    ],
    returns: "boolean" => "Whether fill succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "snapshot",
        action: "tab_snapshot",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Take a DOM snapshot of the target tab and return readable text. Defaults to active tab.",
        params: [
            tab_id: "number", optional, "Target tab ID (defaults to active tab)",
        ],
        returns: "string" => "Human-readable accessibility tree with refIds",
    );
    lua_api!(ctx, tab_table,
        name: "snapshot_text",
        action: "tab_snapshot_text",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Take a DOM snapshot and return readable text (explicit alias). Defaults to active tab.",
        params: [
            tab_id: "number", optional, "Target tab ID (defaults to active tab)",
        ],
        returns: "string" => "Human-readable accessibility tree with refIds",
    );
    lua_api!(ctx, tab_table,
        name: "snapshot_data",
        action: "tab_snapshot_data",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Take a DOM snapshot and return structured data. Defaults to active tab.",
        params: [
            tab_id: "number", optional, "Target tab ID (defaults to active tab)",
        ],
        returns: "table" => "Structured snapshot with nodes, url, title, viewport",
    );
    lua_api!(ctx, tab_table,
    name: "scroll_to",
    action: "tab_scroll_to",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Scroll to an element by refId in the target tab.",
    params: [
    tab_id: "number", required, "Target tab ID",
    ref_id: "number", required, "Element refId from snapshot",
    ],
    returns: "boolean" => "Whether scroll succeeded",
    );
    lua_api!(ctx, tab_table,
    name: "evaluate",
    action: "tab_evaluate",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Evaluate JavaScript in a target tab and return the result.",
    params: [
    tab_id: "number", required, "Target tab ID",
    script: "string", required, "JavaScript code to evaluate",
    ],
    returns: "any" => "Evaluation result",
    );
    lua_api!(ctx, tab_table,
    name: "back",
    action: "tab_back",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Navigate back in a target tab.",
    params: [
    tab_id: "number", required, "Target tab ID",
    ],
    returns: "boolean" => "Whether navigation succeeded",
    );
    lua_api!(ctx, tab_table,
    name: "wait_for_load",
    action: "tab_wait_for_load",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Wait for a tab to finish loading.",
    params: [
    tab_id: "number", required, "Target tab ID",
    timeout: "number", optional, "Timeout in milliseconds (default 30000)",
    ],
    returns: "boolean" => "Whether the tab loaded",
    );
    lua_api!(ctx, tab_table,
        name: "type",
        action: "tab_type",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Type text into an input element by refId in the target tab (appends).",
        params: [
            tab_id: "number", required, "Target tab ID",
            ref_id: "number", required, "Element refId from snapshot",
            text: "string", required, "Text to type",
        ],
        returns: "boolean" => "Whether type succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "press",
        action: "tab_press",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Dispatch a keyboard key press in the target tab.",
        params: [
            tab_id: "number", required, "Target tab ID",
            key: "string", required, "Key to press (e.g. 'Enter', 'Escape')",
        ],
        returns: "boolean" => "Whether press succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "select",
        action: "tab_select",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Select an option in a dropdown by refId in the target tab.",
        params: [
            tab_id: "number", required, "Target tab ID",
            ref_id: "number", required, "Element refId from snapshot",
            value: "string", required, "Option value to select",
        ],
        returns: "boolean" => "Whether select succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "check",
        action: "tab_check",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Toggle a checkbox by refId in the target tab.",
        params: [
            tab_id: "number", required, "Target tab ID",
            ref_id: "number", required, "Element refId from snapshot",
            checked: "boolean", optional, "Desired checked state (default true)",
        ],
        returns: "boolean" => "Whether check succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "hover",
        action: "tab_hover",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Hover over an element by refId in the target tab.",
        params: [
            tab_id: "number", required, "Target tab ID",
            ref_id: "number", required, "Element refId from snapshot",
        ],
        returns: "boolean" => "Whether hover succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "unhover",
        action: "tab_unhover",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Unhover (mouseleave) an element by refId in the target tab.",
        params: [
            tab_id: "number", required, "Target tab ID",
        ],
        returns: "boolean" => "Whether unhover succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "scroll",
        action: "tab_scroll",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Scroll the target tab page.",
        params: [
            tab_id: "number", required, "Target tab ID",
            direction: "string", optional, "Scroll direction: up or down (default down)",
            amount: "number", optional, "Scroll amount in pixels (default 300)",
        ],
        returns: "boolean" => "Whether scroll succeeded",
    );
    lua_api!(ctx, tab_table,
        name: "dblclick",
        action: "tab_dblclick",
        host_state: host_state,
        namespace: "web.tab",
        doc: "Double-click an element by refId in the target tab.",
        params: [
            tab_id: "number", required, "Target tab ID",
            ref_id: "number", required, "Element refId from snapshot",
        ],
        returns: "boolean" => "Whether dblclick succeeded",
    );
    lua_api!(ctx, tab_table,
    name: "fetch",
    action: "tab_fetch",
    host_state: host_state,
    namespace: "web.tab",
    doc: "Perform an HTTP fetch inside a target tab origin.",
    params: [
    tab_id: "number", required, "Target tab ID",
    url: "string", required, "URL to fetch",
    opts: "table | nil", optional, "Options: method, body, headers, timeout",
    ],
    returns: "table" => "{ status, ok, body, headers }",
    );

    tab_table
}

File: crates/extension-lua/js/src/main/runner/tools/tab.ts
// Tab tools — operate on a specific target tab

import { z } from "zod";
import { registerTool } from "../../../shared/tool-registry.js";
import {
  TabBackParamsSchema,
  TabCheckParamsSchema,
  TabClickParamsSchema,
  TabDblClickParamsSchema,
  TabFillParamsSchema,
  TabForwardParamsSchema,
  TabHoverParamsSchema,
  TabPressParamsSchema,
  TabScrollParamsSchema,
  TabSelectParamsSchema,
  TabTypeParamsSchema,
  TabUnhoverParamsSchema,
  TabWaitForLoadParamsSchema,
} from "../../../shared/schemas.js";
import { bridgeToTab } from "../tab/messaging.js";
import { waitForTabLoad } from "../tab/execute.js";
import { throwToolError } from "../../../shared/errors.js";

registerTool({
  action: "tab_click",
  namespace: "tab",
  description: "Click an element in a target tab",
  params: TabClickParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "refId",
      type: "string",
      required: true,
      description: "Element refId from snapshot",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    refId: "Element refId from snapshot",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_click", {
      refId: params.refId,
    });
  },
});

registerTool({
  action: "tab_fill",
  namespace: "tab",
  description: "Fill an input element in a target tab",
  params: TabFillParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "refId",
      type: "string",
      required: true,
      description: "Element refId from snapshot",
    },
    {
      name: "value",
      type: "string",
      required: true,
      description: "Text to fill",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    refId: "Element refId from snapshot",
    value: "Text to fill",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_fill", {
      refId: params.refId,
      value: params.value,
    });
  },
});

registerTool({
  action: "tab_type",
  namespace: "tab",
  description: "Type text into an input in a target tab",
  params: TabTypeParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "refId",
      type: "string",
      required: true,
      description: "Element refId from snapshot",
    },
    {
      name: "text",
      type: "string",
      required: true,
      description: "Text to type",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    refId: "Element refId from snapshot",
    text: "Text to type",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_type", {
      refId: params.refId,
      text: params.text,
    });
  },
});

registerTool({
  action: "tab_press",
  namespace: "tab",
  description: "Press a key in a target tab",
  params: TabPressParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "key",
      type: "string",
      required: true,
      description: "Key to press",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: { tabId: "Target tab ID", key: "Key to press" },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_press", {
      key: params.key,
    });
  },
});
registerTool({
  action: "tab_select",
  namespace: "tab",
  description: "Select an option in a dropdown in a target tab",
  params: TabSelectParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "refId",
      type: "string",
      required: true,
      description: "Element refId from snapshot",
    },
    {
      name: "value",
      type: "string",
      required: true,
      description: "Option value to select",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    refId: "Element refId from snapshot",
    value: "Option value to select",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_select", {
      refId: params.refId,
      value: params.value,
    });
  },
});
registerTool({
  action: "tab_check",
  namespace: "tab",
  description: "Toggle a checkbox in a target tab",
  params: TabCheckParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "refId",
      type: "string",
      required: true,
      description: "Element refId from snapshot",
    },
    {
      name: "checked",
      type: "boolean",
      required: false,
      description: "Desired checked state (default true)",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    refId: "Element refId from snapshot",
    checked: "Desired checked state",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_check", {
      refId: params.refId,
      checked: params.checked,
    });
  },
});
registerTool({
  action: "tab_hover",
  namespace: "tab",
  description: "Hover over an element in a target tab",
  params: TabHoverParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "refId",
      type: "string",
      required: true,
      description: "Element refId from snapshot",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    refId: "Element refId from snapshot",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_hover", {
      refId: params.refId,
    });
  },
});
registerTool({
  action: "tab_unhover",
  namespace: "tab",
  description: "Unhover in a target tab",
  params: TabUnhoverParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: { tabId: "Target tab ID" },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_unhover", {});
  },
});
registerTool({
  action: "tab_scroll",
  namespace: "tab",
  description: "Scroll a target tab",
  params: TabScrollParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "direction",
      type: "string",
      required: false,
      description: "Scroll direction: up or down (default down)",
    },
    {
      name: "amount",
      type: "number",
      required: false,
      description: "Scroll amount in pixels (default 300)",
    },
  ],
  returns: z.boolean(),
  returnDoc: "boolean",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    direction: "Scroll direction",
    amount: "Scroll amount in pixels",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_scroll", {
      direction: params.direction,
      amount: params.amount,
    });
  },
});
registerTool({
  action: "tab_dblclick",
  namespace: "tab",
  description: "Double-click an element in a target tab",
  params: TabDblClickParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "refId",
      type: "string",
      required: true,
      description: "Element refId from snapshot",
    },
  ],
  returns: z.null(),
  returnDoc: "null",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    refId: "Element refId from snapshot",
  },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_dblclick", {
      refId: params.refId,
    });
  },
});
registerTool({
  action: "tab_back",
  namespace: "tab",
  description: "Navigate back in a target tab",
  params: TabBackParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
  ],
  returns: z.boolean(),
  returnDoc: "boolean",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: { tabId: "Target tab ID" },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_back", {});
  },
});
registerTool({
  action: "tab_forward",
  namespace: "tab",
  description: "Navigate forward in a target tab",
  params: TabForwardParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
  ],
  returns: z.boolean(),
  returnDoc: "boolean",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: { tabId: "Target tab ID" },
  handler: async (params) => {
    return bridgeToTab(Number(params.tabId), "tab_forward", {});
  },
});
registerTool({
  action: "tab_wait_for_load",
  namespace: "tab",
  description: "Wait for a target tab to finish loading",
  params: TabWaitForLoadParamsSchema,
  paramTypes: [
    {
      name: "tabId",
      type: "number",
      required: true,
      description: "Target tab ID",
    },
    {
      name: "timeout",
      type: "number",
      required: false,
      description: "Timeout in milliseconds (default 30000)",
    },
  ],
  returns: z.boolean(),
  returnDoc: "true",
  errorCode: "ETAB",
  errorCategory: "tab",
  paramDocs: {
    tabId: "Target tab ID",
    timeout: "Timeout in milliseconds (default 30000)",
  },
  transport: "host_async",
  handler: async (params) => {
    const timeoutMs = Number.isFinite(Number(params.timeout)) && Number(params.timeout) > 0
      ? Number(params.timeout)
      : 30_000;
    const result = await waitForTabLoad(
      Number(params.tabId),
      timeoutMs,
    );
    if (!result.ok) {
      throwToolError(result);
    }
    return result.value;
  },
});

